engine/runtime/src/main/java/org/enso/interpreter/node/callable/dispatch/LoopingCallOptimiserNode.java

```diff
@@ -54,30 +57,43 @@ public static LoopingCallOptimiserNode build() {
    * @param loopNode a cached instance of the loop node used by this node
    * @return the result of executing {@code function} using {@code arguments}
    */
-  @Specialization
+  @Specialization(guards = "warnings == null")
   public Object dispatch(
       Function function,
       CallerInfo callerInfo,
       State state,
       Object[] arguments,
+      Warning[] warnings,
       @Cached(value = "createLoopNode()") LoopNode loopNode) {
     RepeatedCallNode repeatedCallNode = (RepeatedCallNode) loopNode.getRepeatingNode();
     VirtualFrame frame = repeatedCallNode.createFrame();
     repeatedCallNode.setNextCall(frame, function, callerInfo, arguments);
     repeatedCallNode.setState(frame, state);
     loopNode.execute(frame);
-
     return repeatedCallNode.getResult(frame);
   }
 
-  @Specialization(replaces = "dispatch")
+  @Specialization(guards = "warnings != null")
+  public Object dispatchWarnings(
+      Function function,
+      CallerInfo callerInfo,
+      State state,
+      Object[] arguments,
+      Warning[] warnings,
+      @Cached(value = "createLoopNode()") LoopNode loopNode) {
+    Object result = dispatch(function, callerInfo, state, arguments, warnings, loopNode);
+    return WithWarnings.appendTo(EnsoContext.get(this), result, warnings);
```

Comment:
```suggestion
    Object result = dispatch(function, callerInfo, state, arguments, null, loopNode);
    return WithWarnings.appendTo(EnsoContext.get(this), result, warnings);
```

If I read the code right, the `warnings` parameter of `dispatch` is ignored apart from the `Specialization` guard, is that right?

In that case I'd consider passing `null` here, as otherwise it deceptively looks as if the warnings _were indeed propagated_ and it looks as if they were duplicated (passed to function and then also appended). Passing `null` will make it clear that there is only a single path which the warnings follow.

Reply:
refactored it a bit